prov/rxd/src/rxd_cq.c

```diff
@@ -425,20 +443,29 @@ static void rxd_handle_rts(struct rxd_ep *ep, struct rxd_pkt_entry *pkt_entry)
 			"ERROR: Protocol version mismatch with peer\n");
 		return;
 	}
+    rxd_av = rxd_ep_av(ep);
+    fastlock_acquire(&rxd_av->util_av.lock);
 
-	rxd_av = rxd_ep_av(ep);
-	node = ofi_rbmap_find(&rxd_av->rbmap, pkt->source);
+    ret = rxd_av_insert_dg_addr(rxd_av, pkt->source, 
+                &dg_addr, 0, NULL);
+    if (ret){
+        fastlock_release(&rxd_av->util_av.lock);
+        return;
+    }
 
-	if (node) {
-		rxd_addr = (fi_addr_t) node->data;
-	} else {
-		ret = rxd_av_insert_dg_addr(rxd_av, (void *) pkt->source,
-					    &rxd_addr, 0, NULL);
-		if (ret)
-			return;
-	}
+    peer_entry = rxd_get_peer_entry(ep, pkt->source);
+
+    if ((peer_entry->dg_addr == FI_ADDR_UNSPEC)
+        && (peer_entry->peer_addr == FI_ADDR_UNSPEC)) {
+
+        peer_entry->fi_addr = FI_ADDR_UNSPEC;
+        peer_entry->dg_addr = dg_addr;
+        peer_entry->peer_addr = ofi_buf_index(peer_entry);
+        rxd_epaddr_hash_insert(ep, pkt->source, &(peer_entry->peer_addr));
+    }
+    fastlock_release(&rxd_av->util_av.lock);
 
-	if (rxd_send_cts(ep, pkt, rxd_addr)) {
+	if (rxd_send_cts(ep, pkt, peer_entry->peer_addr)) {
```

Comment:
I think the rxd_addr is our local rxd_addr here, not the peer_addr (buf pool index, not peer_addr)

Reply:
fixed this